### ISSUE TITLE
`EarlyDo` - return early from a do-block

### ISSUE THREAD

**Issue by chrisdone** [ghc-proposals/ghc-proposals#503]
## Problem statement

The problem it solves is returning early from a do expression.
```haskell
app :: IO (Either Error String)
app = do
  path <- grabEnv "PATH"?
  putStrLn "Look ma, no lifts!"
  magic <- grabEnv "MAGIC"?
  pure (Right (path ++ magic))
```

## Implementation

The syntax `stmt?` is desugared in this way:

* `do stmt?; next` becomes `do earlyThen stmt next`
* `do pat <- stmt?; next; next2` becomes `do early stmt (\pat -> do next; next2; ...)`

Where early/earlyThen simply return Left if stmt returns Left, or if Right a, they pass the `a` over to the continuation. Nothing clever or special. It’s equivalent to writing

```haskell
do
  result <- m
  case result of
    Left e -> pure (Left e)
    Right a -> f a
```

It could be overloaded to any failure-like type, or not:

```haskell
-- monomorphic version
early :: Monad m => m (Either e a) -> (a -> m (Either e b)) -> m (Either e b)
-- polymorphic version
early :: (Monad m, Early f) => m (f a) -> (a -> m (f b)) -> m (f b)
```

## Why an extension

I’ve been using quite happily this compiler plugin which I wrote a year ago:

https://github.com/inflex-io/early (more details and motivation written there, including FAQs like “why not ExceptT)

I’d like a proper language extension for it. We have arrows and mdo and a bunch other syntactic extensions that I’ve never used for anything, but this one has practical value for me and I’m sure others. My plugin is ok, but has a heavy dependency (ghc-parser-lib), the parser is not 100% accurate and plugins add compile overhead (although not that much).

A factor I didn’t consider, but is true, is that Haskell newbies would be able to use this easily, without having to learn what a monad transformer is. But that’s not really a concern of mine.

I think Haskellers deserve to explore other ways to handle early returning; doing it via clever monad instances seems to have been fully explored.

## Considerations

Technical considerations:

* Should it work for any “Early” type (Either, Maybe, Failure, etc), like [here](https://github.com/inflex-io/early/blob/main/src/Control/Early.hs#L9)
* Or should it be monomorphic? The diff for the two APIs is [here](https://github.com/inflex-io/early/commit/d9638f62ddc6ad09299c4b631d48d582d96f2e16)
* Because ? cannot appear validly at the end of a statement (without parens) in Haskell, this should present no conflicts. But I could be wrong.

If feedback is positive, then I can submit a proposal. And then if accepted, do the GHC language implementation.

## A good summary by Ollie Charles

> I would like to see this. Error handling is an important aspect of programming, and I don't think Haskell has a very good story. Our predominant tools are:
> 
> * Exceptions. These are simple to produce, and also fairly simple to deal with, but they are also essentially untyped (they don't appear in the type of things that throw exceptions). I think this means they work well in the small, but are much more problematic in the large.
> * Monad transformers (e.g., `ExceptT`). This makes it easy to propagate errors, but incur a cost (every bind now has to do pattern matching), and don't combine with all monads (e.g., we can't easily use `ExceptT` with `MonadUnliftIO`)
> * Returning a sum types. This is the approach advocated by this issue, but is tedious for users as they now have to deal with unpacking and error propagation.
> 
> I think there is elegance and simplicity to the third approach - the error appears clearly in the type, and in the return position. It's a shame that actually working in this style is so difficult though, and I like the idea of having a special early-out bind function to ease that pain.
> 
> I'm fairly indifferent to what the actual syntax is. As long as I can just add a few characters to `x <- f a b c` to get error propagation, I'm happy.
> 
> I am sympathetic to concerns about adding extra syntax - Haskell is definitely a heavy language. However, I still feel having canonical syntax for _typed_ error handling in Haskell would be extremely beneficial in the long run.
> 
> The only thing I don't like about this approach is it's very easy to accidentally forget to check for errors for functions that don't return anything:
> 
> ```haskell
> foo :: IO (Either Bang ())
> 
> bar = do
>   ...
>   ...
>   foo -- Whoops, even if this fails we'll continue!
>   foo? -- We should have written this
>   ...
> ```
> 
> However, I think `-Wunused-do-bind` would alert about this.

### Objection: Special syntax for a single type of effect

@Ericson2314 
> @chrisdone OK. Then, this boils down to a difference of opinion.
> 
> I view Rust's `?` as just a weaken form of `!` because they don't have a general notion like `Monad`. `.await` is another example of this. Rust developers by no means like this state of affairs either, see https://without.boats/blog/the-problem-of-effects/ for example. And regular users who don't know what Monads are still wishing they didn't have to, e.g. write things twice for async and regular.
> 
> I would not be a fan of seeing what feels like an ad-hoc per-effect syntax in Haskell when it's no the only option.

### Similarity to case bind 

@tomjaguarpaw 
> ### Similarity to #327
> This seems similar in spirit to #327, though the latter achieves greater generality at the cost of additional syntactic noise. I think what you would write with `EarlyDo` as
> 
> ```haskell
> app = do
>   path <- grabEnv "PATH"?
>   ...
> ```
> 
> you would write in #327 as
> 
> ```haskell
> app = do
>   Right path <- case grabEnv "PATH" of
>      e -> e
>    ...
> ```
> 
> #327 has the benefit that you can distinguish multiple failure cases.

**Comment by tomjaguarpaw** [comment]
### Similarity to #327

This seems similar in spirit to https://github.com/ghc-proposals/ghc-proposals/pull/327, though the latter achieves greater generality at the cost of additional syntactic noise.  I think what you would write with `EarlyDo` as

```haskell
app = do
  path <- grabEnv "PATH"?
  ...
```

you would write in #327 as

```haskell
app = do
  Right path <- case grabEnv "PATH" of
     e -> e
   ...
```

#327 has the benefit that you can distinguish multiple failure cases.

### Type of `earlyThen`

It would be helpful to add the type of `early` and `earlyThen` to the top comment.  From https://github.com/inflex-io/early it seems they are

```haskell
early :: (Monad m, Early f) => m (f a) -> (a -> m (f b)) -> m (f b)
earlyThen :: (Monad m, Early f) => m (f a) -> m (f b) -> m (f b)
```

### Theory

From [the implementation of the `Early` class](https://github.com/inflex-io/early/blob/main/src/Control/Early.hs) I see that the only member is `dispatch :: Applicative m => f a -> (a -> m (f b)) -> m (f b)`. I guess this is equivalent to having an operation `(Applicative m, Early f) => f (m (f a)) -> m (f a)`. If `f` is `Traversable` (as it is for the instances in the `early` package, `Either a` and `Maybe`) then this is strongly related to (perhaps equivalent to?) `(Applicative m, Early f) => m (f (f a))) -> m (f a)`, which is strongly related to (perhaps equivalent to?) `f` being a `Monad`).

Is there any chance that `(Monad f, Traversable f)` is actually the appropriate setting for this feature? Are there any instances of `Early` in the wild that don't satisfy those constraints?

**Comment by chrisdone** [comment]
> ### Similarity to #327
> 
> This seems similar in spirit to #327, though the latter achieves greater generality at the cost of additional syntactic noise. I think what you would write with `EarlyDo` as
> 
> ```haskell
> app = do
>   path <- grabEnv "PATH"?
>   ...
> ```
> 
> you would write in #327 as
> 
> ```haskell
> app = do
>   Right path <- case grabEnv "PATH" of
>      e -> e
>    ...
> ```
> 
> #327 has the benefit that you can distinguish multiple failure cases.

That one has a different intent - as the point of early return is that you don't want to do case analysis - you don't care about the `Left` case. Think [Rust's ? syntax](https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator).

> ### Type of `earlyThen`
> 
> It would be helpful to add the type of `early` and `earlyThen` to the top comment. 

I'll add the monomorphic and polymorphic versions. :+1: 

> ### Theory
> 
> From [the implementation of the `Early` class](https://github.com/inflex-io/early/blob/main/src/Control/Early.hs) I see that the only member is `dispatch :: Applicative m => f a -> (a -> m (f b)) -> m (f b)`. I guess this is equivalent to having an operation `(Applicative m, Early f) => f (m (f a)) -> m (f a)`. If `f` is `Traversable` (as it is for the instances in the `early` package, `Either a` and `Maybe`) then this is strongly related to (perhaps equivalent to?) `(Applicative m, Early f) => m (f (f a))) -> m (f a)`, which is strongly related to (perhaps equivalent to?) `f` being a `Monad`).
>
> Is there any chance that `(Monad f, Traversable f)` is actually the appropriate setting for this feature? Are there any instances of `Early` in the wild that don't satisfy those constraints?

That's a little confusing - the purpose is a class that's isomorphic to case analysis on `Either`. Traversable has instances for tuples and lists, which would be quite strange as a class to use for driving success/failure decision-making.

```haskell
-- monomorphic version
early :: Monad m => m (Either e a) -> (a -> m (Either e b)) -> m (Either e b)
-- polymorphic version
early :: (Monad m, Early f) => m (f a) -> (a -> m (f b)) -> m (f b)
```

**Comment by chrisdone** [comment]
By the way, if there's a class that fits, it would be _nice_ to know about it. Some related classes:

https://hackage.haskell.org/package/selective-0.5/docs/Control-Selective.html#t:Selective
https://hackage.haskell.org/package/pointed-5.0.3/docs/Data-Pointed.html

But it's not that `Traversable`/`Foldable`/`Applicative`/`Selective` came from mathematical theory as such, either, so I wouldn't impose that as a constraint on something like an `Early` class. But others may differ.

I'm not against having an entirely monomorphic `Either` version and then people use `Either ()` rather than `Maybe`. :thinking:

**Comment by ekmett** [comment]
You could almost get there with `QualifiedDo`.  Make a module `Early` with `(>>)` and `(>>=)` defined to be `earlyThen` and `early` respectively and import it qualified. That would cover the cases where everything returns in `IO (Either x y)` at least, but it doesn't handle the mix and match style you use here.

**Comment by chrisdone** [comment]
Right! I explored a [`Try.do` via `QualifedDo`](https://chrisdone.com/posts/try-do/) first before doing the plugin. I realized every statement in the `Try.do` has to return `Either`. I fiddled with some type class trickery to avoid that, but it degraded type inference. Indeed, the lack of being able to simply add a single `?` to a 10-line do statement, or mix-and-match, is a downside.

**Comment by tomjaguarpaw** [comment]
> But it's not that `Traversable`/`Foldable`/`Applicative`/`Selective` came from mathematical theory as such, either, so I wouldn't impose that as a constraint on something like an `Early` class

Well, wherever they came from, `Traversable`, `Applicative` and `Selective` all have laws. I suspect `Early` does too! It would be good to write them down.

**Comment by chrisdone** [comment]
> > But it's not that `Traversable`/`Foldable`/`Applicative`/`Selective` came from mathematical theory as such, either, so I wouldn't impose that as a constraint on something like an `Early` class
> 
> Well, wherever they came from, `Traversable`, `Applicative` and `Selective` all have laws. I suspect `Early` does too! It would be good to write them down.

Absolutely, by all means! :+1:

**Comment by chrisdone** [comment]
@tomjaguarpaw Another way of wording the same class:

```haskell
class Fallible f where
  failure :: (f b -> r) -> (a -> r) -> f a -> r

instance Fallible (Either e) where
  failure left _     (Left e)  = left (Left e)
  failure _    right (Right r) = right r

instance Fallible Maybe where
  failure left _     Nothing  = left Nothing
  failure _    right (Just r) = right r

early :: (Applicative m, Fallible f) => f a -> (a -> m (f b)) -> m (f b)
early fa fab = failure pure fab fa
```

It looks a little bit like a fold, or `either` and `maybe`, but you need the ability to acquire an `f b` in the case of `Left`. That handles the fact that in the middle of your statement list if you have an `Either E X` but the whole do-expression's type is `Either E Y`. You can see that if you look closely at the simpler monomorphic `early`:

```haskell
early :: Monad m => m (Either e a) -> (a -> m (Either e b)) -> m (Either e b)
early m f = do
  r <- m
  case r of
    Left e -> pure (Left e)
    Right x -> f x
```

**Comment by Jashweii** [comment]
> You could almost get there with `QualifiedDo`. Make a module `Early` with `(>>)` and `(>>=)` defined to be `earlyThen` and `early` respectively and import it qualified. That would cover the cases where everything returns in `IO (Either x y)` at least, but it doesn't handle the mix and match style you use here.

I think if QualifiedDo was extended to allow qualifying individual <-s and ;s it would cover EarlyDo and more

**Comment by Ericson2314** [comment]
@chrisdone Have you seen https://idris2.readthedocs.io/en/latest/tutorial/interfaces.html#notation? I only learned about it in https://github.com/ghc-proposals/ghc-proposals/discussions/472#discussioncomment-1769563. It is the proper generalization of Rust's `?`.

**Comment by chrisdone** [comment]
> @chrisdone Have you seen https://idris2.readthedocs.io/en/latest/tutorial/interfaces.html#notation? I only learned about it in [#472 (comment)](https://github.com/ghc-proposals/ghc-proposals/discussions/472#discussioncomment-1769563). It is the proper generalization of Rust's `?`.

I have seen it, and mentioned it in [a gist about applicative syntaxes](https://gist.github.com/chrisdone/8ea87b30d3897a031de3ceeec690815e), but I don’t consider it addressing this problem, nor Rust’s, really. Rust already has early return because it’s imperative. Haskell doesn’t have it, and I don’t personally think it should. Do notation however provides a local, meaningful interpretation of ? syntax because do notation is inherently chaining. The ! syntax is an awkward middle ground between do-notation and effect handlers IMHO.

**Comment by konsumlamm** [comment]
I think the `?` syntax is rather weird for a curried language like Haskell. `path <- grabEnv "PATH"?` makes it seem like the `?` only applies to `"PATH"`. In Rust, this would be `let path = grap_env("PATH")?;`, which makes it obvious what the `?` applies to.

I'm not sure what's a good way to address that though.

**Comment by chrisdone** [comment]
> I think the `?` syntax is rather weird for a curried language like Haskell. `path <- grabEnv "PATH"?` makes it seem like the `?` only applies to `"PATH"`. In Rust, this would be `let path = grap_env("PATH")?;`, which makes it obvious what the `?` applies to.
> 
> I'm not sure what's a good way to address that though.

You could make the same observation for `do f x; y` and say that `do f(x); y` is clearer. Haskell only permits `;` in specific places, so it never seems like it applies to an argument of a function application.

**Comment by chrisdone** [comment]
But your comment highlights the key difference between Rust and Haskell in this. Rust’s functions can return from literally anywhere, following C-like tradition. Haskell’s functions are simple lambdas which have no such capability. You can’t just throw in a `?` and say it’s good, you have to answer questions about ordering and laziness and all sorts of horrible mess. The substitution model of Haskell is lost.
 
As it happens, do-notation gives us a framework for continuation passing style and a very clean translation from the syntax of `?` to underlying the expression. No non-local semantics or jumps or tricks.  Still just pure functions. 

The Rust analogy is helpful, but only up to a point, where the languages differ (imperative vs functional).

**Comment by Ericson2314** [comment]
@chrisdone OK. Then, this boils down to a difference of opinion.

I view Rust's `?` as just a weaken form of `!` because they don't have a general notion like `Monad`. `.await` is another example of this. Rust developers by no means like this state of affairs either, see https://without.boats/blog/the-problem-of-effects/ for example. And regular users who don't know what Monads are still wishing they didn't have to, e.g. write things twice for async and regular.

I would not be a fan of seeing what feels like an ad-hoc per-effect syntax in Haskell when it's no the only option.

**Comment by chrisdone** [comment]
@Ericson2314 
> @chrisdone OK. Then, this boils down to a difference of opinion.
>
> I would not be a fan of seeing what feels like an ad-hoc per-effect syntax in Haskell when it's no the only option.

That’s fine, not everybody has to be a fan. This is a common objection so it’s good to have it voiced. 👌 I will put your complete comment in the top post so that it doesn’t need to be repeated.

**Comment by konsumlamm** [comment]
> > I think the `?` syntax is rather weird for a curried language like Haskell. `path <- grabEnv "PATH"?` makes it seem like the `?` only applies to `"PATH"`. In Rust, this would be `let path = grap_env("PATH")?;`, which makes it obvious what the `?` applies to.
> > I'm not sure what's a good way to address that though.
> 
> You could make the same observation for `do f x; y` and say that `do f(x); y` is clearer. Haskell only permits `;` in specific places, so it never seems like it applies to an argument of a function application.

I view `;` as a separator, while I view `?` as a postfix operator, which is why I don't think this is comparable. In fact, in the Rust example, the `;` also applies to the whole statement (and not just the `grap_env("PATH")?` part. In Rust, `?` can also apply to part of an expression, which doesn't seem to be the case with your proposal, so it isn't actually ambiguous, but it *looks* ambiguous to me.

**Comment by adamse** [comment]
What about idris's "alternatives on binds in do" feature

```
getVal :: m (Maybe Int)
fun :: m _
fun = do
    Just val <- getVal
        | _ => early exit here -- must have the same type as the overall do expression
    do something with val
```

desugaring to case

```
fun = do
    x <- getVal
    case x of
        Just val -> do something with val
        _ => early exit here
```

**Comment by Kleidukos** [comment]
I am against tackling this problem with a new syntax addition. This is typically the kind of stuff that an error monad does well already.

> Haskell newbies would be able to use this easily, without having to learn what a monad transformer is. But that’s not really a concern of mine.

As an educator I will give you my opinion: This is dangerous. We can't as a community face the accusations that Monad Transformers are obscure, while at the same time having GHC enable more ways to do early returns in monadic contexts. 

I am happy that you have a plugin for this, because I think a GHC proposal that comes without code is no better than an oneiric discussion. That being said I'm certain that this feature can live a happy life as a plugin. :)

**Comment by chrisdone** [comment]
> What about idris's "alternatives on binds in do" feature
> 
> ```
> getVal :: m (Maybe Int)
> fun :: m _
> fun = do
>     Just val <- getVal
>         | _ -> early exit here -- must have the same type as the overall do expression
>     do something with val
> ```
> 
> desugaring to case
> 
> ```
> fun = do
>     x <- getVal
>     case x of
>         Just val -> do something with val
>         _ -> early exit here
> ```
> 
> I like this because it doesn't restrict me to using Either in the types.

A similar GHC proposal for the same was mentioned above in the thread: I think that’s a good generalised construct, but it’s heavy syntactically. Do it more than once in a do-expression and it’s a drag.

**Comment by ocharles** [comment]
I would like to see this. Error handling is an important aspect of programming, and I don't think Haskell has a very good story. Our predominant tools are:

* Exceptions. These are simple to produce, and also fairly simple to deal with, but they are also essentially untyped (they don't appear in the type of things that throw exceptions).
* Monad transformers (e.g., `ExceptT`). This makes it easy to propagate errors, but incur a cost (every bind now has to do pattern matching), and don't combine with all monads (e.g., we can't easily use `ExceptT` with `MonadUnliftIO`)
* Returning a sum types. This is the approach advocated by this issue, but is tedious for users as they now have to deal with all the unpacking and error propagation.

I think there is some elegant simplicity to the third approach - the error appears clearly in the type, and in the return position. It's a shame that actually working in this style is so difficult though, and I like the idea of having a special early-out bind function to ease that pain.

I'm fairly indifferent to what the actual syntax is. As long as I can just add a few characters to `x <- f a b c` to get error propagation, I'm happy.

I am sympathetic to concerns about adding extra syntax - Haskell is definitely a heavy language. However, I still feel having canonical syntax for *typed* error handling in Haskell would be extremely beneficial in the long run.

**Comment by chrisdone** [comment]
> I am against tackling this problem with a new syntax addition. This is typically the kind of stuff that an error monad does well already.

Fair enough. FWIW. Addressed this in the early GitHub link, along with accompanying posts: ExceptT doesn’t do it well.

> > Haskell newbies would be able to use this easily, without having to learn what a monad transformer is. But that’s not really a concern of mine.
> 
> As an educator I will give you my opinion: This is dangerous. We can't as a community face the accusations that Monad Transformers are obscure, while at the same time having GHC enable more ways to do early returns in monadic contexts.

I’d leave speaking for the community to the community itself.

**Comment by chrisdone** [comment]
@ocharles Thanks for the very nicely written summary.

I also think the unused do binds would be needed to make this properly reliable, good point to raise. I should add that to the top description. ☝🏻

**Comment by chrisdone** [comment]
@tomjaguarpaw I added your observation of similarity to case bind to the objections section. 👌

**Comment by NorfairKing** [comment]
While I like the contents of the proposal, I don't like the parts of this proposal that are not described: How will the ecosystem be upgraded?

There's an unknown unknown amount of tooling that will need to be upgraded to support adding even more syntax to the already enormous language:
* HLS
* Ormolu/Fourmolu
* Styling-haskell
* haskell-src-exts

**Comment by Ericson2314** [comment]
@ocharles 

> Monad transformers (e.g., ExceptT). This makes it easy to propagate errors, but incur a cost (every bind now has to do pattern matching), and don't combine with all monads (e.g., we can't easily use ExceptT with MonadUnliftIO)

This proposal still makes an `m (Either a b)`, so I don't think these are addressed.

**Comment by Gabriel439** [comment]
I'm against this because you can achieve the same thing using `ExceptT`:

```haskell
{-# LANGUAGE BlockArguments #-}

module Example where

import Control.Monad.Trans.Class
import Control.Monad.Trans.Except

data Error = Error

grabEnv :: String -> ExceptT Error IO String
grabEnv = undefined

app :: ExceptT Error IO String
app = do
  path <- grabEnv "PATH"
  lift (putStrLn "Here is a lift")
  magic <- grabEnv "MAGIC"
  return (path ++ magic)
```

The proposed `EarlyDo` solution also doesn't play well with the existing ecosystem.  Every utility in the Haskell ecosystem that is parametric over the `Monad` class would need to be forked to create an `Early`-aware version in order to interoperate with this convention / idiom (including, but not limited to, the `Control.Monad` / `Control.Applicative` / `Data.Traversable` / `Data.Foldable` modules in `base`).

For example, `traverse` works with `ExceptT` but does not work with actions of type `IO (Either e a)` (unless you wrap them in `ExceptT`), so you now have to define an `Early`-aware version of `traverse`

As a concrete example of why we want to play nice with existing abstractions, suppose you removed the `putStrLn` from the example so that it becomes:

```
app :: ExceptT Error IO String
app = do
  path <- grabEnv "PATH"
  magic <- grabEnv "MAGIC"
  return (path ++ magic)
```

… if you use `ExceptT` for error-handling then you could transform that example to:

```haskell
app = liftA2 (++) (grabEnv "PATH") (grabEnv "MAGIC")
```

… but that transformation would not work if you used actions of type `IO (Either a b)`.  You'd have to define an `Early`-aware version of `liftA2`

From my point of view, here is the cost-benefit analysis:

* Benefits:
  * You don't have to wrap non-failing actions in `lift` or `liftIO`
* Costs:
  * You have to wrap the return value in `Right`
  * Requires forking every utility that is parametric over `Applicative` / `Monad` (e.g. `traverse`)
  * More syntax/semantics that new users have to learn
  * Adds more complexity to the language standard / implementation

In other words, I don't believe this proposed extensions carries its own weight and adds more complexity than it reduces

**Comment by ocharles** [comment]
@Ericson2314 I don't understand your comment - could you elaborate?

**Comment by chrisdone** [comment]
@Gabriel439 For a complete cost-benefit list you should address `MonadUnliftIO`, which, as stated, is the primary reason to reject `ExceptT`.

**Comment by gbaz** [comment]
Imho the answer is simple -- use literally any other way of lifting handling through a stack except for the only one too restrictive to allow `Maybe` and `Either`.

**Comment by chrisdone** [comment]
> Imho the answer is simple -- use literally any other way of lifting handling through a stack except for the only one too restrictive to allow `Maybe` and `Either`. (Otherwise your transformer stack is incredibly restricted anyway).

Thanks, but I’ll let @Gabriel439 address this explicitly, in case she has a more nuanced answer than “use something else.”

> (edit: and if you're in a stack over IO, you already have a canonical way to perform an early exit anyway -- throw an exception!)

You write this as a foregone conclusion but it may surprise you to hear that some Haskellers find using untyped exceptions for everything problematic!

**Comment by Gabriel439** [comment]
It's not clear to me how actions of type `IO (Either e r)` work better with `MonadUnliftIO`.  As a concrete example, suppose you wanted to use the `UnliftIO.Async.withAsync` utility, which has the following example:

```haskell
withAsync :: MonadUnliftIO m => m a -> (a -> m b) -> m b
```

… if you were to specialize the argument to a value of type `IO (Either e r)`, then the specialized type would be:

```haskell
withAsync :: IO (Either e r) -> (Either e r -> IO b) -> IO b
```

… which runs the risk of the error value being accidentally forwarded to the forked thread instead of being handled within the current thread

More generally, I don't feel that the language design should be significantly influenced by the `MonadUnliftIO` class since it's a pretty ad-hoc class that only has instances for `IO` / `Identity` / `ReaderT`.  It's nowhere near as principled and general-purpose as, say, `Monad` / `Applicative` / `Functor` and shouldn't carry the same weight in these design discussions

**Comment by Ericson2314** [comment]
Beyond just feeling that we shouldn't have more sugar unless it is really compelling (why I am also tepid about my own https://github.com/ghc-proposals/ghc-proposals/pull/327) error handling is in Haskell today is far too discordant with all the effects vs MTL noise. That makes it an especially bad time to add error handling sugar.

> You write this as a foregone conclusion but it may surprise you to hear that some Haskellers find using untyped exceptions for everything problematic!

I would *much* rather see a proposal on providing a way to ban synchronous exceptions throughout the entire program.

**Comment by chrisdone** [comment]
> Beyond just feeling that we shouldn't have more sugar unless it is really compelling (why I am also tepid about my own #327) error handling is in Haskell today is far too discordant with all the effects vs MTL noise. That makes it an especially bad time to add error handling sugar.

Or a great time, if you’re sceptical of effect systems and want something different. ☝🏻

**Comment by chrisdone** [comment]
> which runs the risk of the error value being accidentally forwarded to the forked thread instead of being handled within the current thread

The point of withAsync is to run a thread on the side and pass the result to the action, but you’re describing this behaviour as a “risk.” It would be no accident. Perhaps you’re presuming an interpretation of error handling that looks like exceptions?

> More generally, I don't feel that the language design should be significantly influenced by the MonadUnliftIO class since it's a pretty ad-hoc class that only has instances

Ok, I thought there might be something more nuanced than “pretty ad hoc”, but I see @gbaz’s guess was accurate. It was worth checking, however.

**Comment by tomjaguarpaw** [comment]
> Or a great time, if you’re sceptical of effect systems and want something different.

Such folks already have something different: your GHC plugin.  Like effect systems, which are opt-in and whose carrying cost is borne by their maintainers, your plugin is opt-in and its carrying cost is borne by you. Unless I am much mistaken, no one in this thread is saying you shouldn't want something different to what Haskell's ecosystem already provides, and you shouldn't use your plugin. Rather, I think objection to this proposal stems from a belief that adding syntax baked in to GHC imposes a carrying cost on the GHC maintainers which ultimately affects all users of GHC, and imposes a carrying cost on all members of the community who have to add it to their mental repository of what GHC provides.

**Comment by chrisdone** [comment]
Sure, that’s the point of this GitHub issue; I think others are speaking for themselves well enough. 👌

**Comment by Ericson2314** [comment]
> Or a great time, if you’re sceptical of effect systems and want something different. ☝🏻

OK, I am too! But I don't think this would help tamp down on effects mania. I already mentioned banning synchronous exceptions, but I would also propose these perf changes:

 - Make sure the MTL change to separate using to handling exceptions goes through.
 - Monomorphizing/templating type quantification, so we can can do unboxed MTL without `-fexpose-all-unfoldings`, praying, and waiting, and shatter any notion of effect systems being fast
 - "multi return functions" i.e. allow pushing multiple return pointers, so we don't need to branch on `Either`.

On the ergonomics side the answers are less clear to me but

 - Like I side, better macros to allow library-level exploring do notation like things. You wrote https://chrisdone.com/posts/haskell-doesnt-have-macros/ as I think you agree :)
   - I think it is also easier than it sounds. In particular, https://github.com/haskell/happy/issues/235 (see liked paper) should allow us to unfuck layout, after which I think we can make a Haskell equivalent of "token trees".
 - I also think we overuse Monad in general (!) --- too many people are doing imperative programming rather than trying to better understand the structure of their coterminating program. Hopefully we will see new developments in an arrows / FRP / self-adjusting computation vein. Macros also helps with this too! (Arrows is definitely not well-baked enough to be in GHC.)
 
   If all that sounds too galaxy brained in general, a syntax for *communative* monads (Reader-like things) would be nice. And Idris's `!` is good for that because there is no implicit ordering to leverage.
 - Once all those things happen, I think it will be far less clear what sort of error handling syntax we want, because there will be such semantic diversity of what people are doing. And that's good!

**Comment by gbaz** [comment]
> You write this as a foregone conclusion but it may surprise you to hear that some Haskellers find using untyped exceptions for everything problematic!

I don't advocate using exceptions for everything. _However_ if you are using `MonadUnliftIO`, then the purpose of that is almost always _precisely_ to lift exception handlers through a monad transformer stack over IO. Which means, that if you are using that, then you already _have_ made the decision to use exceptions for some things (rather than catching and wrapping them immediately in every IO action that throws them). So the particular issue of "how do I use Either for exceptions if I already have made a decision to use a particular different idiom for exceptions" seems a bit odd.

If you instead make the decision to not use exceptions, and instead use `Either` or something similar pervasively, then the idiomatic solution is a transformer stack.

Also, to be clear, I was not speculating what @Gabriel439 might answer. I was just giving my own answer.

**Comment by ocharles** [comment]
> Which means, that if you are using that, then you already have made the decision to use exceptions for some things

Perhaps pedantic, but those of us who reach for `MonadUnliftIO` do so out of necessity, not choice. Many _other_ bits of code that we don't want to reinvent decided to use exceptions, not us!  And even if we did wrap all of that code in `tryAny`, we still have to `bracket` and `mask` around asynchronous exceptions. I'm not sure there really is a decision that can be made here.

**Comment by gbaz** [comment]
`bracket` is a perfect example of why mixing exception models is a bad idea. Suppose you `bracket` some code and then attempt to "early exit" with this idiom from that bracketed code. This won't "early exit" from the enclosing do block but give back an `Either` that needs to be manually propagated back up through the main do block. Now, there, _is_ an idiom which would do that automatically. That idiom is monad transformers with any more expressive handler lifting choice than `MonadUnliftIO`. 

`catch` is another good example. Catch in such an idiom would handle exceptions, but not trigger on a `Left` -- so use of this idiom under a catch would not "catch" the exception condition causing the early exit. So now, you need to write two handlers for exception conditions, not one. First, the handler for the actual exception, second, the handler for the `Left`.

**Comment by nomeata** [comment]
Thanks for raising this, though; even if a proposal doesn’t make it all the way to “accepted” it is useful, I’d say.

**Comment by Icelandjack** [comment]
*Early* is similar to *Traversable* where the continuation returns the whole structure back. This means it inherits the techniques (and issues) with T.

```Haskell
for      :: Traversable t => Applicative f => t a -> (a -> f b)     -> f (t b)
dispatch ::       Early t => Applicative f => t a -> (a -> f (t b)) -> f (t b)
```

Analogously to *sequenceA* we now have (by Yoneda lowering):

```haskell
unroll :: Early t => Applicative f => t (f (t a)) -> f (t a)
unroll = (`dispatch` id)
```

Similar to *{fmap,foldMap}Default* we get *via{Identity,Const}*. *viaIdentity* seems to be an implementation of Monadic bind, and *viaConst* implements *flip foldMap*.

```haskell
viaIdentity :: forall t a b. Early t => t a -> (a -> t b) -> t b
viaIdentity = coerce do dispatch @t @Identity @a @b

viaConst :: forall t m a. Early t => Monoid m => t a -> (a -> m) -> m
viaConst = coerce do dispatch @t @(Const m) @a @()
````

This also means *Early* [like *Traversable* cannot be derived by coercion](https://ryanglscott.github.io/2018/06/22/quantifiedconstraints-and-the-trouble-with-traversable/), because the role of the *Applicative f* must be (conservatively) assumed nominal.

If the interface were a (Co)Yoneda transformed *dispatch*, an equivalent interface given the `Functor` constraint, it could be coerce-derived. Now *dispatch* becomes a regular function, which changes the interface.

```haskell
class Functor t => Early t where
  {-# minimal dispatchOne | dispatchTwo #-}
  dispatchOne :: Applicative f => (bs -> t b) -> (t b -> res) -> t a -> (a -> f bs) -> f res
  dispatchOne pre post as g = dispatchTwo post as \f -> fmap (f . pre) . g

  dispatchTwo :: Applicative f => (t b -> res) -> t a -> (forall ts. (t b -> ts) -> a -> f ts) -> f res
  dispatchTwo post as f = post <$> dispatch as (f id)

dispatch :: Early t => Applicative f => t a -> (a -> f (t b)) -> f (t b)
dispatch = dispatchOne id id
```

One solution to this is to separate the (coerce-friendly) implementation from the interface we expose to the user. I wrote about this idea

* [Type class backend: how to evolve with class](https://www.reddit.com/r/haskell/comments/pqhivm/type_class_backend_how_to_evolve_with_class/]

We can also use the [Representation Theorem for Second-Order Functionals](https://arxiv.org/abs/1402.1699) (theorem 3.5), and derive an alternative representation of `dispatch`.

```haskell
  forall f. Applicative f => t a -> (a -> f (t b)) -> f (t b)
= forall f. Applicative f => t a -> (forall bs. a -> (t b -> bs) -> f bs) -> f (t b)
= forall f. Applicative f => t a -> (forall bs. Búð t a b bs -> f bs) -> f (t b)
= forall f. Applicative f => t a -> (Búð t a b ~> f) -> f (t b)
= forall f. Applicative f => t a -> (FreeApplicative (Búð t a b) ~> f) -> f (t b)
= t a -> FreeApplicative (Búð t a b) (t b)

data Búð t a b bs = Búð a (t b -> bs)
```
